### PR TITLE
The Simulator!

### PR DESCRIPTION
**Pull request by snarayan21**
## Description of changes:
Adding a simulator for streaming.

## Issue #, if available:
Simulator story: https://mosaicml.atlassian.net/browse/STR-123

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [x] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -93,6 +93,14 @@
     'sphinx-tabs==3.4.1',
 ]
 
+extra_deps['simulator'] = [
```
Comment (karan6181) [inline setup.py:96]: Can you restrict the package version to a major version with `<` operator?

File: simulation/core/create_index.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Create a dataset index file from input parameters."""
+
+import os.path
+import sys
+
+sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
```
Comment (karan6181) [inline simulation/core/create_index.py:9]: Why do we need this?

File: simulation/core/create_index.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            'hashes': [],
+            'samples': samples_per_shard,
+            'size_limit': avg_raw_shard_size,
+            'version': 2,
+            'zip_data': None,
+            'compression': None
+        }
+        if avg_zip_shard_size is not None:
+            shard_data['zip_data'] = {
+                'basename': 'shard.' + str(shard_id) + '.mds.zstd',
```
Comment (karan6181) [inline simulation/core/create_index.py:61]: Any reason why compression extension is hard coded?

File: simulation/core/create_index.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    index_data['shards'] = shards_list
+
+    # Try making the directory for the stream's index.json file
+    foldername = get_random_foldername() + '_indexcreated'
+    try:
+        os.mkdir(foldername)
+    except FileExistsError:
+        print('Folder already exists, trying again...')
+        foldername = get_random_foldername()
+        os.mkdir(foldername)
```
Comment (karan6181) [inline simulation/core/create_index.py:77]: What happens if there is a collision here? Do we want to loop through this logic until we find a unique directory? Maybe try 100/1000 times or so?

File: simulation/core/create_index.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    with open(foldername + '/index.json', 'w') as f:
+        json.dump(index_data, f)
+        f.close()
```
Comment (karan6181) [inline simulation/core/create_index.py:81]: You don't need to call `f.close()` since that's handled in `with` call.

File: simulation/core/create_index.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    # Try making the directory for the stream's index.json file
+    foldername = get_random_foldername() + '_indexcreated'
+    try:
+        os.mkdir(foldername)
+    except FileExistsError:
+        print('Folder already exists, trying again...')
+        foldername = get_random_foldername()
+        os.mkdir(foldername)
+
+    with open(foldername + '/index.json', 'w') as f:
```
Comment (karan6181) [inline simulation/core/create_index.py:79]: For `index.json`, you can use the method [get_index_basename()](https://github.com/mosaicml/streaming/blob/6b638518294f7338f7bb54e9bb529bfad8e5fa50/streaming/base/format/index.py#L9)

File: simulation/core/create_index.py (hunk 2 above)
Comment (snarayan21) [reply simulation/core/create_index.py:9]: Oh yeah I put this here by mistake. Within the `core` directory this isn't needed, but when we want to import functionality from the `core` directory in the `interfaces` or `testing` directories then it's necessary. Otherwise we get import errors.

File: simulation/core/create_index.py (hunk 3 above)
Comment (snarayan21) [reply simulation/core/create_index.py:61]: Removed all hardcoding from shard basenames in `raw_data` and `zip_data` when creating this index file. All the fields have to be present because otherwise StreamingDataset will throw an error if the index file is not structured correctly, but no hardcoding / making fake shard names is actually needed so I've removed that now.

File: simulation/core/create_index.py (hunk 4 above)
Comment (snarayan21) [reply simulation/core/create_index.py:77]: So the folders get generated and deleted right after they are processed, during `SimulationDataset` instantiation. So there are only a few of these folders that exist at a time. The `get_random_foldername()` function returns a random alphanumeric string of length 16. characters can be lowercase, uppercase, and numeric, so that's 26+26+10 = 62 choices for each character, so overall 62^16 ~= 4.7e28 possible filenames, so the collision probability is super super low. I just added this `FileExistsError` in the very slim off-chance there's a collision, but if you'd want I can increase the number of tries?

File: simulation/core/create_index.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        shards_list.append(shard_data)
+
+    index_data['shards'] = shards_list
+
+    # Try making the directory for the stream's index.json file
+    foldername = get_random_foldername() + '_indexcreated'
+    try:
+        os.mkdir(foldername)
+    except FileExistsError:
+        print('Folder already exists, trying again...')
```
Comment (karan6181) [inline simulation/core/create_index.py:68]: Should we convert all the `print()` statements to `logger.info()`?

File: simulation/core/node_tracker.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+from sortedcollections import OrderedSet
+
+from streaming.base.spanner import Spanner
+
+
+class NodeTracker():
+    """Tracker for node information during simulation.
+
+    Args:
+        workers (int): The number of workers.
```
Comment (karan6181) [inline simulation/core/node_tracker.py:21]: Are these CPU workers? The `num_workers` parameter in Dataloader?

File: simulation/core/create_index.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+import json
+import os
+import random
+import string
+from typing import Optional
+
+from streaming.base.format import get_index_basename
+
+
+def get_random_foldername():
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply simulation/core/create_index.py:15]: Just tried using this and there's problems with using the `TemporaryDirectory()` because it falls out of scope once the function completes. Even returning the `TemporaryDirectory` from the function causes it to fall out of scope, meaning that it will get cleaned up before accessed in `SimulationDataset`. Because the index file creation functionality is separated from the instantiation of the StreamingDataset, I had to use actual folders instead of temp dirs.

File: simulation/core/create_index.py (hunk 7 above)
Comment (snarayan21) [reply simulation/core/create_index.py:68]: changed all `print` statements to `logger.info()` or `logger.warning()` except for those in `sim_cli.py` and `sim_script.py` since it looks nicer being kept as print statements there.

File: simulation/core/main.py [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+    for _ in range(physical_nodes):
+        nodes.append(
+            NodeTracker(workers, devices, predownload, device_batch_size, total_shards,
+                        cache_limit))
+
+    # Time for the global batch is just device batch size * time per sample.
+    # We assume all devices process their microbatch perfectly in parallel.
+    avg_batch_process_time = device_batch_size * time_per_sample
+
+    notification_batches = int(batches_per_epoch) / 20
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply simulation/core/main.py:94]: it's just so we log batches completed so far 20 times every simulation run. So the notification batch interval is 1/20th the total number of steps in the simulation.

File: simulation/core/node_tracker.py (hunk 8 above)
Comment (snarayan21) [reply simulation/core/node_tracker.py:21]: yeah it's the same.

File: simulation/core/node_tracker.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+    def evict_shard(self) -> int:
+        """Evict a shard.
+
+        Returns:
+            int: The evicted shard.
+        """
+        evicted_shard = self.shards.popLRU()
+        return evicted_shard
+
+    def evict_until_satisfied(self, incoming_shard_size: int, raw_shard_sizes: NDArray[np.int64]):
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply simulation/core/node_tracker.py:112]: No, this happens for the overall node. But even in the real simulation dataset, shard evictions happen through the worker that is trying to download a shard that pushes the cache usage past the cache limit. So the overall effect is essentially the same, right?

File: simulation/core/shuffle_quality.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        device_batch_size (int): The batch size per device.
+        samples_per_shard (int): Average number of samples per shard.
+        remove_padding (bool): Whether to remove padding samples. Defaults to ``False``.
+
+    Returns:
+        Tuple[NDArray, NDArray, NDArray]: The partition, in order, the
+            sizes of each shard, and the mapping of sample id to shard id.
+    """
+    num_samples = epoch_size
+    if num_samples > 100000000:
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply simulation/core/shuffle_quality.py:72]: oh cool I didn't know you could do that in regular python code!

File: simulation/core/sim_time.py [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+                             r'|'.join([fr'{time_unit.value}' for time_unit in TimeUnit]) + r')$',
+                             flags=re.IGNORECASE)
+
+TValue = TypeVar('TValue', int, float)
+
+
+class Time(Generic[TValue]):
+    """Time represents static durations of training time in terms of a :class:`TimeUnit` enum.
+
+    See the :doc:`Time Guide </trainer/time>` for more details on tracking time during training.
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply simulation/core/sim_time.py:44]: Oh yeah this is straight from composer. changed the docstring to be simpler and just refer to composer docs.

File: simulation/requirements.txt [hunk 14]
```diff
@@ -0,0 +1,10 @@
+altair==5.1.1
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply simulation/requirements.txt:1]: oh whoops i'll delete this. it's the same and the requirements have already been moved to setup.py.

File: simulation/interfaces/widgets.py [hunk 15]
```diff
@@ -0,0 +1,446 @@
...
+def param_inputs(component: DeltaGenerator, input_params: dict, defaults: dict = {}):
+    """Define parameter input area widget.
+
+    Args:
+        component (DeltaGenerator): Streamlit component to write to.
+        input_params (dict): Dictionary to store input parameters.
+        defaults (dict): Dictionary of default values for input params. Defaults to empty dict, {}.
+    """
+    # split the input column component into left, middle, and right sub columns.
+    colL, colM, colR = component.columns(3)
```
Comment (karan6181) [inline simulation/interfaces/widgets.py:178]: nit: underscore naming convention?

File: simulation/interfaces/sim_ui.py [hunk 16]
```diff
@@ -0,0 +1,397 @@
...
+    immediate_throughput_data = []
+    network_data = []
+    steps = []
+    time_to_first_batch = 0
+    futures = []
+    shuffle_quality_graphed = False
+    # Initialize min_cache_limit to be 0. Will be replaced by the simulated value.
+    min_cache_limit = 0
+    # Define partial function to pass to executor map for simulation.
+    with ProcessPoolExecutor(max_workers=8) as executor:
```
Comment (karan6181) [inline simulation/interfaces/sim_ui.py:81]: nit: Do we need `max_workers=8`? I think it is by default set as a number of logical CPU cores in your system.

File: simulation/interfaces/sim_ui.py [hunk 17]
```diff
@@ -0,0 +1,397 @@
...
+            # Use multiprocessing to get the shuffle quality results.
+            canonical_nodes = input_params['canonical_nodes']
+            physical_nodes = input_params['physical_nodes']
+            devices = input_params['devices']
+            workers = input_params['workers']
+            device_batch_size = input_params['device_batch_size']
+            shuffle_block_size = number_abbrev_to_int(input_params['shuffle_block_size'])
+            samples_per_shard = dataset.get_avg_samples_per_shard()
+            epoch_size = dataset.get_epoch_size()
+            if epoch_size > 100000000:
```
Comment (karan6181) [inline simulation/interfaces/sim_ui.py:95]: nit: `100000000` -> `100_000_000` ?

File: simulation/interfaces/sim_ui.py [hunk 18]
```diff
@@ -0,0 +1,397 @@
...
+    # Initialize min_cache_limit to be 0. Will be replaced by the simulated value.
+    min_cache_limit = 0
+    # Define partial function to pass to executor map for simulation.
+    with ProcessPoolExecutor(max_workers=8) as executor:
+        # Submit shuffle quality job to executor.
+        if shuffle_quality:
+            col1.write('Starting shuffle quality analysis...')
+            input_params = st.session_state['input_params']
+            # Use multiprocessing to get the shuffle quality results.
+            canonical_nodes = input_params['canonical_nodes']
```
Comment (karan6181) [inline simulation/interfaces/sim_ui.py:87]: Question: Is there a case where a key is not present in the dictionary?

File: simulation/core/yaml_processing.py [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+    max_duration = ensure_time(max_duration, TimeUnit.EPOCH)
+    time_unit = max_duration.unit
+    if time_unit != TimeUnit.EPOCH and time_unit != TimeUnit.BATCH:
+        raise ValueError('Simulator currently only supports max_duration in epochs or batches.')
+
+    # convert train_dataset to dict, if it isn't already
+    if isinstance(train_dataset, DictConfig):
+        train_dataset = om.to_container(train_dataset)
+
+    assert isinstance(workers, int), 'workers must be an integer.'
```
Comment (karan6181) [inline simulation/core/yaml_processing.py:110]: I would recommend raising a `ValueError` instead of an `assert` statement in the source code.

File: simulation/README.md [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+- Find optimal run configurations
+- Debug issues with underperforming runs
+- Better understand the impact of different configurations
+
+## 🚀 Getting Started
+Run the following to install simulator-specific dependencies, if they don't already exist:
+```
+pip install --upgrade "mosaicml-streaming[simulator]"
+```
+Then, simply run `simulator` in your command line to open the Web UI and get simulating!
```
Comment (karan6181) [inline simulation/README.md:13]: Should we rename this to `sd_simulator` or `sd-simulator` or something else other than `simulator`? Wondering, if there is a name collision with other application or not?

Comment (snarayan21) [reply simulation/README.md:13]: google search for a `simulator` command on linux and mac doesn't turn up any results...so we should probably be fine? I think the simpler the better right

File: simulation/interfaces/sim_ui.py (hunk 16 above)
Comment (snarayan21) [reply simulation/interfaces/sim_ui.py:81]: ooh nice. checked and confirmed you're right!

File: simulation/interfaces/sim_ui.py (hunk 18 above)
Comment (snarayan21) [reply simulation/interfaces/sim_ui.py:87]: nope, this dict is always initialized in `get_input_params_initial()`

File: simulation/core/yaml_processing.py (hunk 19 above)
Comment (snarayan21) [reply simulation/core/yaml_processing.py:110]: done for all lines!

File: simulation/README.md (hunk 20 above)
Comment (snarayan21) [reply simulation/README.md:13]: discussed and resolved. if issues, will change later!

**Review by karan6181 — APPROVED** [review]
LGTM. Thank You!
